src/core/models.py

```diff
@@ -0,0 +1,283 @@
+        self,
+        relation: Relation | None,
+        data_interface: Data,
+        component: Application,
+        substrate: SUBSTRATES,
+        local_app: Application | None = None,
+        password: str = "",
+        endpoints: str = "",
+        tls: bool = False,
+    ):
+        super().__init__(relation, data_interface, component, substrate)
+        self.app = component
+        self._password = password
+        self._endpoints = endpoints
+        self._tls = tls
+        self._local_app = local_app
+
+    @override
+    def update(self, items: dict[str, str]) -> None:
+        """Overridden update to allow for same interface, but writing to local app bag."""
+        if not self.relation or not self._local_app:
+            return
+
+        self.data_interface.update_relation_data(self.relation.id, items)
+
+    @property
+    def username(self) -> Optional[str]:
+        """The generated username for the client application."""
+        # Until we settle user credential questions we statically return 'kibanaserver'
+        # return self.relation_data.get("username")
+        return "kibanaserver"
+
+    @property
+    def password(self) -> Optional[str]:
+        """The generated password for the client application."""
+        return self.relation_data.get("password")
+
+    @property
+    def endpoints(self) -> List[str]:
+        """Connection endpoints for the client application to connect with."""
+        endpoints_str = self.relation_data.get("endpoints")
+        return endpoints_str.split(",") if endpoints_str else []
+
+    @property
+    def tls(self) -> bool:
+        """Flag to confirm whether or not is TLS enabled.
+
+        Returns:
+            String of either 'enabled' or 'disabled'
+        """
+        return self._tls
+
+    @property
+    def tls_ca(self) -> Optional[str]:
+        """The CA cert in case TLS is enabled.
+
+        Returns:
+            String of either 'enabled' or 'disabled'
+        """
+        return self.relation_data.get("tls-ca")
+
+
+class ODCluster(StateBase):
+    """State collection metadata for the charm application."""
+
+    def __init__(
+        self,
+        relation: Relation | None,
+        data_interface: Data,
+        component: Application,
+        substrate: SUBSTRATES,
+    ):
+        super().__init__(relation, data_interface, component, substrate)
+        self.app = component
+
+    @property
+    def internal_user_credentials(self) -> dict[str, str]:
+        """The passwords for the internal quorum and super users.
+
+        Returns:
+            Dict of key username, value password
+        """
+        return {
+            user: password
+            for user in CHARM_USERS
+            if (password := self.relation_data.get(f"{user}-password"))
+        }
+
+    # -- TLS --
+
+    @property
+    def tls(self) -> bool:
+        """Flag to check if TLS is enabled for the cluster."""
+        return bool(self.relation_data.get("tls", ""))
```

Comment:
The issue here is that it makes TLS setup depend on both `peer` and `tls` relations for no real gain.

Reply:
Is it such an issue really?

Are we particularly keen on having TLS relations before a peer relation is established?

To me this doesn't feel more than a strong corener-case, that I'm not sure if it should be addressed even.

BTW peer TLS (i.e. databag usage typically) will be reviewed in the [final cleanup](https://warthogs.atlassian.net/browse/DPE-3954). As indeed we may not need things I've copied from ZK. 

Here I'd rather like to focus on having a base (fully functional and tested) version of the charm, even if we may apply changes/refactors on some of the debateable behaviors later).